src/download.rs

```diff
@@ -0,0 +1,300 @@
+//! # Download large messages manually.
+
+use anyhow::{anyhow, Result};
+use deltachat_derive::{FromSql, ToSql};
+use serde::{Deserialize, Serialize};
+
+use crate::config::Config;
+use crate::context::Context;
+use crate::imap::{Imap, ImapActionResult};
+use crate::job::{self, Action, Job, Status};
+use crate::message::{Message, MsgId};
+use crate::param::Params;
+use crate::{job_try, EventType};
+use std::cmp::max;
+
+/// Download limits should not be used below `MIN_DOWNLOAD_LIMIT`.
+///
+/// Some messages as non-delivery-reports (NDN) or read-receipts (MDN)
+/// need to be downloaded completely to handle them correctly,
+/// eg. to assign them to the correct chat.
+/// As these messages are typically small,
+/// they're catched by `MIN_DOWNLOAD_LIMIT`.
+const MIN_DOWNLOAD_LIMIT: u32 = 32768;
+
+/// If `delete_server_after` is set to small timeouts (eg. "at once"),
+/// the user might have no chance to actually download a message.
+/// `MIN_DELETE_SERVER_AFTER` increases the timeout in this case.
+pub const MIN_DELETE_SERVER_AFTER: i64 = 48 * 60 * 60;
+
+#[derive(
+    Debug,
+    Display,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    FromPrimitive,
+    ToPrimitive,
+    FromSql,
+    ToSql,
+    Serialize,
+    Deserialize,
+)]
+#[repr(u32)]
+pub enum DownloadState {
+    Done = 0,
+    Available = 10,
+    Failure = 20,
+    InProgress = 1000,
+}
+
+impl Default for DownloadState {
+    fn default() -> Self {
+        DownloadState::Done
+    }
+}
+
+impl Context {
+    // Gets validated download limit or 0 for "no limit".
+    pub(crate) async fn get_download_limit(&self) -> Result<u32> {
```

Comment:
`Result<Option<u32>>` looks better to me than using `0` as a special value meaning "no limit".

Reply:
changed, the usage of `get_download_limit()` in `imap.rs` is now a bit more verbose, but imho also better readable:

https://github.com/deltachat/deltachat-core-rust/pull/2631/commits/9365dbd77f3a6fe3469a1a969350bbb56ea81b92#diff-e574a8b5cb145a5a0d3829cd5290a736d64938f8b1af48de1d1686e7e319c123R716